// generator-sp2010/app/index.js
'use strict';
var util = require('util');
var path = require('path');
var yeoman = require('yeoman-generator');


module.exports = Sp2010Generator;

function Sp2010Generator(args, options, config) {
  yeoman.generators.Base.apply(this, arguments);

  this.indexFile = this.readFileAsString(path.join(this.sourceRoot(), 'index.html'));
  this.mainJsFile = '';

  this.on('end', function () {
    console.log('\nI\'m all done. Just run ' + 'npm install && bower install'.bold.yellow + ' to install the required dependencies.');
    console.log('To start up the web server run ' + 'grunt'.bold.yellow);
  });

  this.pkg = JSON.parse(this.readFileAsString(path.join(__dirname, '../package.json')));
}

util.inherits(Sp2010Generator, yeoman.generators.NamedBase);

Sp2010Generator.prototype.askFor = function askFor() {
  var cb = this.async();
  var welcome =
  '\n     _--_     '.grey +
  '\n    /   /     '.grey +
  '\n   -~  -~     '.grey +
  '\n  ('.white + 'o'.blue + ') ('.white + 'o'.blue + ')     '.white +
  '\n   ||  | /    '.grey +
  '\n   ||  ||     '.grey +
  '\n   \\\\__/|     '.grey +
  '\n    \\___/     \n'.grey;

  console.log(welcome);
  console.log('It looks like you\'re trying to make a web app. Would you like some help with that?\n');

  var prompts = [{
    name: 'description',
    message: 'How would you describe your project?',
    default: 'A SharePoint 2010 REST project.'
  },
  {
    name: 'port',
    message: 'What\'s a good port for us to run the REST server on?',
    default: 8080
  },
  {
    name: 'includeDataTable',
    message: 'Include a boilerplate dataTables example?',
    default: 'Y/n'
  },
  {
    name: 'includeGallery',
    message: 'Include a boilerplate picture gallery example?',
    default: 'Y/n'
  }];

  this.prompt(prompts, function (err, props) {
    if (err) {
      return this.emit('error', err);
    }

    this.description = props.description;
    this.port = parseInt(props.port, 10);
    this.includeDataTable = (/y/i).test(props.includeDataTable);
    this.includeGallery = (/y/i).test(props.includeGallery);

    cb();
  }.bind(this));
};

Sp2010Generator.prototype.gruntfile = function gruntfile() {
  this.template('_Gruntfile.js', 'Gruntfile.js');
};

Sp2010Generator.prototype.packageJSON = function packageJSON() {
  this.template('_package.json', 'package.json');
};

Sp2010Generator.prototype.componentJSON = function componentJSON() {
  this.copy('.bowerrc', '.bowerrc');
  this.template('_component.json', 'component.json');
};

Sp2010Generator.prototype.projectfiles = function projectfiles() {
  this.copy('editorconfig', '.editorconfig');
  this.copy('jshintrc', '.jshintrc');
};

Sp2010Generator.prototype.writeIndex = function writeIndex() {
  var scripts = [];
  var styles = [];
  var contentText = '';

  if(this.includeDataTable || this.includeGallery){
    scripts.push('components/jquery/jquery.min.js');
  }
  if(this.includeDataTable) {
    this.copy('../../datatables/Browsers.json', 'app/lists/Browsers.json' );
    contentText += this.readFileAsString(path.join(this.sourceRoot(), '../../datatables/index.html'));
    this.mainJsFile += this.readFileAsString(path.join(this.sourceRoot(), '../../datatables/js/main.js'));
    scripts.push('components/datatables/media/js/jquery.dataTables.js');
    styles.push('components/datatables/media/css/jquery.dataTables.css');
  }
  if(this.includeGallery) {
    this.copy('../../picture-gallery/Pictures.json', 'app/lists/Pictures.json' );

    this.copy('../../picture-gallery/Pictures/bike.jpg', 'app/Pictures/bike.jpg' );
    this.copy('../../picture-gallery/Pictures/_t/bike_jpg.jpg', 'app/Pictures/_t/bike_jpg.jpg' );
    this.copy('../../picture-gallery/Pictures/cage.jpg', 'app/Pictures/cage.jpg' );
    this.copy('../../picture-gallery/Pictures/_t/cage_jpg.jpg', 'app/Pictures/_t/cage_jpg.jpg' );
    this.copy('../../picture-gallery/Pictures/city.jpg', 'app/Pictures/city.jpg' );
    this.copy('../../picture-gallery/Pictures/_t/city_jpg.jpg', 'app/Pictures/_t/city_jpg.jpg' );
    this.copy('../../picture-gallery/Pictures/path.jpg', 'app/Pictures/path.jpg' );
    this.copy('../../picture-gallery/Pictures/_t/path_jpg.jpg', 'app/Pictures/_t/path_jpg.jpg' );
    this.copy('../../picture-gallery/Pictures/stirfry.jpg', 'app/Pictures/stirfry.jpg' );
    this.copy('../../picture-gallery/Pictures/_t/stirfry_jpg.jpg', 'app/Pictures/_t/stirfry_jpg.jpg' );
    this.copy('../../picture-gallery/Pictures/surf.jpg', 'app/Pictures/surf.jpg' );
    this.copy('../../picture-gallery/Pictures/_t/surf_jpg.jpg', 'app/Pictures/_t/surf_jpg.jpg' );

    this.copy('../../picture-gallery/images/arrows.png', 'app/images/arrows.png' );
    this.copy('../../picture-gallery/images/arrows-small.png', 'app/images/arrows-small.png' );
    this.copy('../../picture-gallery/images/white-opacity-80.png', 'app/images/white-opacity-80.png' );
    this.copy('../../picture-gallery/images/black-opacity-40.png', 'app/images/black-opacity-40.png' );

    contentText += this.readFileAsString(path.join(this.sourceRoot(), '../../picture-gallery/index.html'));
    this.mainJsFile += this.readFileAsString(path.join(this.sourceRoot(), '../../picture-gallery/js/main.js'));
    scripts.push('components/handlebars.js/dist/handlebars.js');
    this.copy('../../picture-gallery/js/jquery.aw-showcase.min.js', 'app/scripts/jquery.aw-showcase.min.js' );
    scripts.push('scripts/jquery.aw-showcase.min.js');

    this.copy('../../picture-gallery/css/awkward.css', 'app/css/awkward.css' );
    styles.push('css/awkward.css');
  }
  scripts.push('scripts/main.js');

  this.indexFile = this.appendStyles(this.indexFile, 'css/styles.js', styles);
  this.indexFile = this.appendScripts(this.indexFile, 'scripts/plugins.js', scripts);

  // 'cause 4 spaces is just too many spaces Yeoman
  this.indexFile = this.indexFile.replace(/        <script src/g, '    <script src');

  this.indexFile = this.indexFile.replace('<body>', '  <body>\n' + contentText);
};

Sp2010Generator.prototype.app = function app() {
  this.mkdir('app');
  this.mkdir('app/lists');
  this.mkdir('app/public');
  this.mkdir('app/templates');
  this.write('app/index.html', this.indexFile);
  this.write('app/scripts/main.js', this.mainJsFile);
};

module.exports = Sp2010Generator;
